data/tutorials/language/0lg_08_basic_datatypes.md

```diff
@@ -0,0 +1,686 @@
+type ('a, 'b) pair = Pair of 'a * 'b
+```
+
+Where `(int, bool) pair` would be written `int * bool` and `Pair (42, true)` would be written `(42, true)`. From the developer's perspective, everything happens as if such a type would be declared for every possible product shape. This is what allows pattern matching on products.
+
+Even integers and floats can be seen as enumerated-like variant types, with many constructors and funky syntactic sugar. This is what allows pattern matching on those types.
+
+In the end, the only type construction that does not reduce to a variant is the function arrow type. No pattern matching on functions.
+
+#### User-Defined Polymorphic
+
+Here is an example of a variant type that combines constructors with data and without data, polymorphism, and recursion:
+```ocaml
+# type 'a tree =
+  | Leaf
+  | Node of 'a * 'a tree * 'a tree;;
+type 'a tree = Leaf | Node of 'a * 'a tree * 'a tree
+```
+
+It can be used to represent arbitrarily labelled binary trees. Using pattern matching, here is how the map function can be defined in this type:
+```ocaml
+# let rec map f = function
+  | Leaf -> Leaf
+  | Node (x, lft, rht) -> Node (f x, map f lft, map f rht);;
+val map : ('a -> 'b) -> 'a tree -> 'b tree = <fun>
+```
+
+**Remark**: OCaml has something called _Polymorphic Variants_. Although the types `option`, `list`, and `tree` are variants and polymorphic, they aren't polymorphic variants. They are type-parametrised variants. Among the functional programming community, the word “polymorphism” is used loosely whenever anything can be applied to various types. We stick to this usage and say the variants in this section are polymorphic. OCaml polymorphic variants are [covered in another tutorial](https://v2.ocaml.org/learn/tutorials/labels.html).
+
+### Records
+
+Records are like tuples in that several values are bundled together. In a tuple, components are identified by their position in the corresponding product type. They are either first, second, third, or at some other position. In a record, each component has a name. That's why record types must be declared before being used.
+
+For instance, here is the definition of a record type meant to partially represent a Dungeons & Dragons character class.
+```ocaml
+# type character = {
+  name : string;
+  level : int;
+  race : string;
+  class_type : character_class;
+  alignment : firmnness * rectitude;
+  armor_class : int;
+};;
+type character = {
+  name : string;
+  level : int;
+  race : string;
+  class_type : character_class;
+  alignment : firmness * rectitude;
+  armor_class : int;
+}
+```
+This is using the types `character_class` and `character_alignment` defined earlier. Values of type `character` carry the same data as inhabitants of this product: `string * int * string * character_class * character_alignment * int`.
+
+Access to the fields is done using the dot notation. Here is an example:
+```ocaml
+# let ghorghor_bey = {
```

Comment:
I get errors from here through li 651.

e.g., `Error: Unbound record field name`, then lots of things like `Error: Unbound value ghorghor_bey` (assumedly since the first didn't work).

Reply:
Can't reproduce this one.

Reply:
Fixed with the definition solution from yesterday... in a continuous UTop session